### ISSUE TITLE
ContinuousDeployment not working

### ISSUE THREAD

**Issue by pascalberger** [GitTools/GitVersion#728]
I just wanted to use GitVersion for some continuous deployment scenario, where I would require the Major.Minor.Patch version for each commit to be different. I'm using GitHubFlow and set `mode: ContinuousDeployment` in `GitVersionConfig.yaml`.

In the following example I'm 29 committs ahead of the last tag `1.0.0` on the master branch and would expect to get version `1.0.29` but getting `1.0.1` instead:

```
{
  "Major":1,
  "Minor":"0",
  "Patch":1,
  "PreReleaseTag":"BuildPubProfileOptional.29",
  "PreReleaseTagWithDash":"-BuildPubProfileOptional.29",
  "BuildMetaData":"",
  "BuildMetaDataPadded":"",
  "FullBuildMetaData":"Branch.BuildPubProfileOptional.Sha.ce7118b30467f748ddaedf091cc00971040c4631",
  "MajorMinorPatch":"1.0.1",
  "SemVer":"1.0.1-BuildPubProfileOptional.29",
  "LegacySemVer":"1.0.1-BuildPubProfileOpt29",
  "LegacySemVerPadded":"1.0.1-BuildPubProfileO0029",
  "AssemblySemVer":"1.0.1.0",
  "FullSemVer":"1.0.1-BuildPubProfileOptional.29",
  "InformationalVersion":"1.0.1-BuildPubProfileOptional.29+Branch.BuildPubProfileOptional.Sha.ce7118b30467f748ddaedf091cc00971040c4631",
  "BranchName":"BuildPubProfileOptional",
  "Sha":"ce7118b30467f748ddaedf091cc00971040c4631",
  "NuGetVersionV2":"1.0.1-buildpubprofileo0029",
  "NuGetVersion":"1.0.1-buildpubprofileo0029",
  "CommitDate":"2015-12-02"
}
```

GitVersion /output buildserver returns:

```
INFO [12/02/15 16:24:40:74] Working directory: c:\Repos\sharepoint-build-tasks-private
INFO [12/02/15 16:24:40:78] Project root is: c:\Repos\sharepoint-build-tasks-private
INFO [12/02/15 16:24:40:96] Using latest commit on specified branch
INFO [12/02/15 16:24:40:98] Running against branch: BuildPubProfileOptional (ce7118b30467f748ddaedf091cc00971040c4631)
INFO [12/02/15 16:24:41:01] Begin: Calculating base versions
  INFO [12/02/15 16:24:41:04] Fallback base version: 0.1.0 with commit count source e6829c47f8a3c289b3ac2fc112bcd7434be30c26
  INFO [12/02/15 16:24:41:18] Git tag '1.0.0': 1.0.0 with commit count source 2740d2709b890fd0d3a4d07374b005f7f9673a55
  INFO [12/02/15 16:24:41:19] Git tag '1.0.0': 1.0.0 with commit count source 2740d2709b890fd0d3a4d07374b005f7f9673a55
  INFO [12/02/15 16:24:41:23] Found multiple base versions which will produce the same SemVer (1.0.1), taking oldest source for commit counting (Git tag '1.0.0')
  INFO [12/02/15 16:24:41:23] Base version used: Git tag '1.0.0': 1.0.0 with commit count source 2740d2709b890fd0d3a4d07374b005f7f9673a55
INFO [12/02/15 16:24:41:23] End: Calculating base versions (Took: 221.16ms)
INFO [12/02/15 16:24:41:24] Using branch name to calculate version tag
INFO [12/02/15 16:24:41:25] 29 commits found between 2740d2709b890fd0d3a4d07374b005f7f9673a55 and ce7118b30467f748ddaedf091cc00971040c4631
```

Repo is this one: https://github.com/pascalberger/sharepoint-build-tasks

**Comment by pushasha** [comment]
I am having the same issue. Even tried committing directly to master with similar results -- only the metadata is being incremented, not the version number. Mode is set to `ContinuousDeployment` on all branches. Below is the `gitversion` output for two sequential commits to master.
***

Commit no. 7:
```
{
  "Major":2,
  "Minor":1,
  "Patch":1,
  "PreReleaseTag":"",
  "PreReleaseTagWithDash":"",
  "BuildMetaData":7,
  "BuildMetaDataPadded":"0007",
  "FullBuildMetaData":"7.Branch.master.Sha.c5587196ef1e35b727a467015e9d13ebd40d0adf",
  "MajorMinorPatch":"2.1.1",
  "SemVer":"2.1.1",
  "LegacySemVer":"2.1.1",
  "LegacySemVerPadded":"2.1.1",
  "AssemblySemVer":"2.1.1.0",
  "FullSemVer":"2.1.1+7",
  "InformationalVersion":"2.1.1+7.Branch.master.Sha.c5587196ef1e35b727a467015e9d13ebd40d0adf",
  "BranchName":"master",
  "Sha":"c5587196ef1e35b727a467015e9d13ebd40d0adf",
  "NuGetVersionV2":"2.1.1",
  "NuGetVersion":"2.1.1",
  "CommitDate":"2016-01-13"
}
```
GitVersion /output buildserver: 
```
INFO [01/13/16 15:19:24:11] Working directory: C:\redacted\gitversiontest\TestApplication
INFO [01/13/16 15:19:24:14] Project root is: C:\redacted\gitversiontest
INFO [01/13/16 15:19:24:19] Using latest commit on specified branch
INFO [01/13/16 15:19:24:20] Running against branch: master (c5587196ef1e35b727a467015e9d13ebd40d0adf)
INFO [01/13/16 15:19:24:22] Begin: Calculating base versions
  INFO [01/13/16 15:19:24:24] Fallback base version: 0.1.0 with commit count source 0be625493dee18c66795ebf9c934a51f3e565f1b
  INFO [01/13/16 15:19:24:30] Git tag 'v2.1.0': 2.1.0 with commit count source a1002059c347540803d1ef417043e09a07a9de1c
  INFO [01/13/16 15:19:24:30] Git tag 'v2.0.0': 2.0.0 with commit count source 5edf94f577f9a29786e9737bae936109a78c7068
  INFO [01/13/16 15:19:24:31] Git tag '1.0.0': 1.0.0 with commit count source 38208ed55ebc5a207a18183e43f86d1a6de4ddda
  INFO [01/13/16 15:19:24:33] Base version used: Git tag 'v2.1.0': 2.1.0 with commit count source a1002059c347540803d1ef417043e09a07a9de1c
INFO [01/13/16 15:19:24:34] End: Calculating base versions (Took: 113.00ms)
INFO [01/13/16 15:19:24:34] 7 commits found between a1002059c347540803d1ef417043e09a07a9de1c and c5587196ef1e35b727a467015e9d13ebd40d0adf

```

Commit no. 8:
```
{
  "Major":2,
  "Minor":1,
  "Patch":1,
  "PreReleaseTag":"",
  "PreReleaseTagWithDash":"",
  "BuildMetaData":8,
  "BuildMetaDataPadded":"0008",
  "FullBuildMetaData":"8.Branch.master.Sha.d35fca99dc6089441b546fff49d9e8af4fe89c3c",
  "MajorMinorPatch":"2.1.1",
  "SemVer":"2.1.1",
  "LegacySemVer":"2.1.1",
  "LegacySemVerPadded":"2.1.1",
  "AssemblySemVer":"2.1.1.0",
  "FullSemVer":"2.1.1+8",
  "InformationalVersion":"2.1.1+8.Branch.master.Sha.d35fca99dc6089441b546fff49d9e8af4fe89c3c",
  "BranchName":"master",
  "Sha":"d35fca99dc6089441b546fff49d9e8af4fe89c3c",
  "NuGetVersionV2":"2.1.1",
  "NuGetVersion":"2.1.1",
  "CommitDate":"2016-01-13"
}
```
GitVersion /output buildserver: 
```
INFO [01/13/16 15:28:57:69] Working directory: C:\redacted\gitversiontest\TestApplication
INFO [01/13/16 15:28:57:72] Project root is: C:\redacted\gitversiontest
INFO [01/13/16 15:28:57:77] Using latest commit on specified branch
INFO [01/13/16 15:28:57:78] Running against branch: master (d35fca99dc6089441b546fff49d9e8af4fe89c3c)
INFO [01/13/16 15:28:57:80] Begin: Calculating base versions
  INFO [01/13/16 15:28:57:83] Fallback base version: 0.1.0 with commit count source 0be625493dee18c66795ebf9c934a51f3e565f1b
  INFO [01/13/16 15:28:57:88] Git tag 'v2.1.0': 2.1.0 with commit count source a1002059c347540803d1ef417043e09a07a9de1c
  INFO [01/13/16 15:28:57:89] Git tag 'v2.0.0': 2.0.0 with commit count source 5edf94f577f9a29786e9737bae936109a78c7068
  INFO [01/13/16 15:28:57:89] Git tag '1.0.0': 1.0.0 with commit count source 38208ed55ebc5a207a18183e43f86d1a6de4ddda
  INFO [01/13/16 15:28:57:92] Base version used: Git tag 'v2.1.0': 2.1.0 with commit count source a1002059c347540803d1ef417043e09a07a9de1c
INFO [01/13/16 15:28:57:92] End: Calculating base versions (Took: 118.00ms)
INFO [01/13/16 15:28:57:93] 8 commits found between a1002059c347540803d1ef417043e09a07a9de1c and d35fca99dc6089441b546fff49d9e8af4fe89c3c`

**Comment by JakeGinnivan** [comment]
Looks like we don't have a test, anyone want to submit a PR for this?

The test can be based off https://github.com/GitTools/GitVersion/blob/master/src/GitVersionCore.Tests/IntegrationTests/DevelopScenarios.cs#L122 but put into `MasterScenarios`.

The desired behavior is that we *always* have a tag on master. (I think the default fallback is `ci`). so master builds are `2.1.0-beta8` in your example.

**Comment by robdmoore** [comment]
Why would you want a tag on master?

**Comment by nshenoy** [comment]
I just submitted PR #733 to add tests as @JakeGinnivan requested. However, I do find the behavior a bit unexpected. For the continuous delivery scenario, if we start with a tag in master of `1.0.0` and then make a commit, the version becomes `1.0.1+1`. So it bumps the patch _and_ adds the build metadata. For the continuous deployment scenario, the same actions result in a pre-release version of `1.0.1-ci.1` (i.e. the patch is bumped and a prerelease identifier is added). Shouldn't ContinuousDeployment mode just increment the patch with each commit and use the SemVer release identifier semantics (that is, no pre-release version) ?

**Comment by JakeGinnivan** [comment]
@robdmoore Say you are using octopus, you want to only bump minor when you release to a test environment, not the continuous deployment environment.
 
Nuget doesn’t allow you to have multiple packages of the same version. As part of octopus deployment to non-dev environments you would drop the tag from the version to get a stable version and tag during the first promotion.

@nshenoy 

> SemVer release identifier semantics

Nope, see http://gitversion.readthedocs.org/en/latest/more-info/incrementing-per-commit which links to a blog post.

In short, GitVersion is a SemVer tool, not a generic versioning tool. And the goal is to *not* have multiple patch version bumps for a single release. https://github.com/GitTools/GitVersion/issues/621 is discussing a new mode which will be possibly better.

**Comment by robdmoore** [comment]
That's right. We've had this chat before lol.

**Comment by JakeGinnivan** [comment]
I found a commit counting bug when I did #744

Could one of you grab https://github.com/JakeGinnivan/GitVersion/tree/BumpDevelopWhenReleaseBranchCreated and check to see if the commit counting is fixed? If not a failing hand would be awesome

**Comment by asbjornu** [comment]
Is this problem still unfixed?

**Comment by asbjornu** [comment]
Closing due to inactivity.
